### ISSUE TITLE
Hitting "left of '=' is not an l-value" when trying to write to UAV.

### ISSUE THREAD

**Issue by zbethel** [shader-slang/slang#415]
The issue reproduces with the compute shader from Falcor on the latest:

https://github.com/NVIDIAGameWorks/Falcor/blob/55f943ef35ee752f368fc14fbc6352818af48367/Samples/Core/ComputeShader/Data/compute.hlsl

Any write to a UAV of the pattern:

uavTexture[coord] = foo;

fails with this error.

**Comment by zbethel** [comment]
I'm seeing it happen in the latest master release.

I also am able to get around the issue by explicitly swizzling:

m_uav[coords].rgba = color; // this works.
m_uav[coords] = color; // this doesn't.

I also attempted to debug and it looks like the left and right arguments of the assignment expression are swapped. I wasn't able to make much headway though because the parser is pretty opaque.

**Comment by tfoleyNV** [comment]
Okay. I’ll take a look at it. Thanks for identifying this issue.

Since you have a workaround for now, is it okay if you don’t get a fix until Monday? That is probably the earliest I’ll be able to get this debugged and implemented.

**Comment by nbenty** [comment]
This is resolved in Falcor's development branch, which will be merged into master in a few weeks.
There's a workaround and I prefer not to do a validation cycle for `master` at the moment.

**Comment by zbethel** [comment]
It's fine, I'm not blocked by it.

**Comment by tfoleyNV** [comment]
@zbethel Okay, I've triaged this a bit and I think I know what is going on. I'm sorry I spoke incorrectly about what Slang `v0.9.12` supported.

Just to confirm, it seems like you are either:
* writing your shader code in a file with the `.slang` suffix, or
* writing your code in a `.hlsl` file and then running on Vulkan

If that is the case, then I'm sorry to say that neither of those is an officially supported use case for Falcor users. Some of the built-in samples use `.slang` files just because it is easier for the Falcor development team, but in general users should be writing `.hlsl` files for D3D12 and `.glsl` files for Vulkan.

The best fix for now will be for you to either write `.hlsl` or `.glsl` shaders, depending on the API you are targetting. A future Falcor release will officially support `.slang` shaders from users, but we aren't quite ready for that today.

**Comment by zbethel** [comment]
To confirm, it's the former.

I'm actually not using this for Falcor, it's for an internal project. My main reason for using it is that I need a quick API that lets me reflect HLSL and group them into the ParameterBlocks.

One potentially solution here could be to run my parameter block include shaders through the .slang compiler, expose them via HLSL function getters, and then write the bulk of the rest of my code via HLSL.

**Comment by tfoleyNV** [comment]
Ah, okay. If you are using Slang directly then I can try to push a new release for you with a fix.

**Comment by zbethel** [comment]
I have a question. Is it possible to run the slang compiler on certain #includes, and then run my main shader as HLSL?

I would like to use the ParameterBlock<> semantics to groups of resources, but I'm finding that the slang compiler itself is not compatible with the HLSL code I currently have.

My thinking is to create .slang include files which define the ParameterBlock<> definition, and then utilize getters GetObject_ModelToWorld(...) for example. 

Is this possible? I would need to be able to run the slang compiler on .slang includes, but run the hlsl compiler on the rest.

**Comment by tfoleyNV** [comment]
Release/tag `v0.9.13` should fix the original issue in this request.

To your other question, the releases of Slang so far support a `-no-checking` flag that turns off most semantic checks (but is only respected on `.hlsl` input files). This is what Falcor currently uses. That might work for your use case.

Looking ahead, though, Slang is dropping support for that flag, and the plan is to only support “full” compilation with all-Slang input. The goal is to eventually support as much of HLSL as people need in practice, but I can understand why the current level of support could be frustrating.

If you want to contact me directly (either DM `@TangentVector` on Twitter, or contact `tfoley` at `nvidia.com`) I’d be interested to hear about your use case and whether it aligns with where the Slang project is headed.

**Comment by tfoleyNV** [comment]
Oh yeah, you had a more specific question:

> For clarity, my specific request is whether I'm able to run the slang code-generator over a file without having to specify an entry point, as this would allow me to process my dependent include files separately from the main source file.

You can currently feed code into Slang without an entry point, and get reflection information back, but in that case you aren’t going to get any output HLSL (output code is per-entry-point). I don’t know if that is useful to you or not.

I suspect that for the scenario you are describing, you probably want go write your entry points as HLSL files that `import` Slang files with your parameter declarations, and then feed those HLSL files into Slang with the `-no-checking` flag. That should give you more or less what you want.

We’ve dropped that whole model in our top-of-tree code, though, because it was adding a lot of implementation complexity, and it would never work for cross-compilation. The future of Slang is going to be its new IR-based compilation flow, and we will treat any HLSL we can’t understand as a bug (which isn’t promising that we will *fix* every such bug).

**Comment by zbethel** [comment]
Great, thanks for the clarification! :)

**Comment by tfoleyNV** [comment]
I think the original problem here is addressed, so I'm going to close this issue.
